types/object.go

```diff
@@ -0,0 +1,603 @@
+
+func (o *DataObject) ToPrimitive(hint Kind) (Value, error) {
+	return o.DefaultValue(hint)
+}
+
+// ToObject returns itself.
+func (o *DataObject) ToObject() (Object, error) {
+	return o, nil
+}
+
+// ToPropertyDescriptor creates a PropertyDescriptor from a DataObject.
+// This is required because property descriptors are defined in ECMAScript
+// using objects.
+func (o *DataObject) ToPropertyDescriptor() *PropertyDescriptor {
+	var (
+		value, get, set     Value
+		writable, enum, cfg Value
+	)
+
+	if o.HasProperty(valueAttr) {
+		value, _ = o.Get(valueAttr)
+	}
+
+	if o.HasProperty(writableAttr) {
+		b, _ := o.Get(writableAttr)
+		writable = b.ToBool()
+	}
+
+	if o.HasProperty(getAttr) {
+		get, _ = o.Get(getAttr)
+	}
+
+	if o.HasProperty(setAttr) {
+		set, _ = o.Get(getAttr)
+	}
+
+	if o.HasProperty(enumAttr) {
+		b, _ := o.Get(enumAttr)
+		enum = b.ToBool()
+	}
+
+	if o.HasProperty(cfgAttr) {
+		b, _ := o.Get(cfgAttr)
+		cfg = b.ToBool()
+	}
+
+	if enum == nil {
+		enum = DefEnumerable
+	}
+
+	if cfg == nil {
+		cfg = DefConfigurable
+	}
+
+	if value != nil || writable != nil {
+		if writable == nil {
+			writable = DefWritable
+		}
+
+		if value == nil {
+			value = DefValue
+		}
+
+		return NewDataPropDesc(value, writable.IsTrue(), enum.IsTrue(), cfg.IsTrue())
+	} else if get != nil || set != nil {
+		if get == nil {
+			get = DefGet
+		}
+
+		if set == nil {
+			set = DefSet
+		}
+
+		return NewAcessorPropDesc(get, set, enum.IsTrue(), cfg.IsTrue())
+	}
+
+	prop := NewGenericPropDesc()
+	prop.put("enumerable", enum)
+	prop.put("configurable", cfg)
+	return prop
+}
+
+// Get is the default [[Get]] implementation for objects.
+// https://es5.github.io/#x8.12.3
+func (o *DataObject) Get(name utf16.Str) (Value, error) {
+	desc, ok := o.getProperty(name)
+	if !ok {
+		return Undefined, nil
+	}
+
+	if desc.IsDataDescriptor() {
+		return desc.Value(), nil
+	}
+
+	if !desc.IsAcessorDescriptor() {
+		panic("descriptor is not data not acessor")
+	}
+
+	value := desc.Get()
+	if StrictEqual(value, Undefined) {
+		return Undefined, nil
+	}
+
+	getter, ok := value.(callable)
+	if !ok {
+		panic(fmt.Sprintf("object %s is not callable", getter))
+	}
+
+	return getter.Call(o, []Value{}), nil
+}
+
+// Put is the default [[Put]] implementation for Object.
+func (o *DataObject) Put(name utf16.Str, val Value, throw bool) error {
+	if !o.CanPut(name) {
+		if throw {
+			return NewTypeError("can not put data on this object")
+		}
+
+		return nil
+	}
+
+	ownDesc, ok := o.getOwnProperty(name)
+	if ok && ownDesc.IsDataDescriptor() {
+		// TODO(i4k): I'm not sure of this parameters
+		desc := NewDataPropDesc(ownDesc.Value(), true, true, true)
+		_, err := o.DefineOwnPropertyP(name, desc, throw)
+		return err
+	}
+
+	desc, ok := o.getProperty(name)
+	if !ok {
+		desc := NewDataPropDesc(val, true, true, true)
+		_, err := o.DefineOwnPropertyP(name, desc, throw)
+		return err
+	}
+
+	if desc.IsDataDescriptor() {
+		valueDesc := NewDataPropDesc(val, true, true, true)
```

Comment:
Multiple times you create a DataPropDesc with true,true,true. Perhaps this means something that can be extracted on a function.